### DISCUSSION TITLE
Unable to set stage quality with new SDK

### DISCUSSION THREAD

**Discussion by Bimeche** [airsdk/Adobe-Runtime-Support#1729] [Q&A]
Hello,

I began working with the Harman SDK some time ago, but I only saw that I can't change the stage quality anymore when using it, compared to when I was using AIR 32 or previous versions.
![image](https://user-images.githubusercontent.com/16176517/156208444-2422b62b-f0e2-4fa8-a70b-91aae3402e6d.png)

Is it a bug or is it designed like that ? Is there another way to set stage quality ? Or what is it I am doing wrong ?

Thanks for the help !

**Comment by ajwfrost** [comment]
It's been restricted to just "high" and "best" for quite a while from my understanding.... I'm not entirely sure why this restriction was added though, see https://github.com/airsdk/Adobe-Runtime-Support/issues/1545#issuecomment-1013251027

We had been looking to add an opt-out for this e.g. `<allowLowQuality>` setting within the `<initialWindow>` value of the application descriptor, please let us know your thoughts on that?

thanks

**Reply by Bimeche** [reply]
Adding a `<allowLowQuality>` setting would be one way of doing it I suppose. You could either have this, or a list of enabled qualities if someone doesn't want to enable one quality level for a reason or another (not my case, but just thinking about all possibilities).

To explain a bit more to why I raised the issue :
I used, until very recently, Adobe AIR 29.0, and there I could set the quality level to LOW or MEDIUM. After reading the Adobe doc, and your comment on the issue you linked, I don't know how it works, but it just does, and that leaves me quite surprised. And when i switch to the new SDK (we use version 33.1.1.674 for info), it stops working. So I don't really know what's happening here, but it's really weird.
I tried using `BitmapData.drawWithQuality(..., "low")` to force it, but it's no use, it stays in High quality.

**Reply by ajwfrost** [reply]
To clarify: on desktop platforms, the quality is set via a string using the Stage object, and this is restricted to high/best. With the below code using AIR 32, I get HIGH/HIGH/BEST/BEST/HIGH/HIGH; I just also checked with AIR 17 and get the same...
```
		trace(stage.quality);
		stage.quality = StageQuality.LOW;
		trace(stage.quality);
		stage.quality = StageQuality.BEST;
		trace(stage.quality);
		stage.quality = StageQuality.MEDIUM;
		trace(stage.quality);
		stage.quality = StageQuality.HIGH;
		trace(stage.quality);
		stage.quality = StageQuality.HIGH_16X16_LINEAR;
		trace(stage.quality);
```
If you're able to check with AIR 29 and see if you are getting the quality coming back as LOW? (on a desktop platform....)

I'm not 100% sure what you're actually trying to do, but if I draw a sprite to a bitmap data and then scale that, it comes out without any smoothing (other than you'd expect from the initial draw):
```
			var s : Sprite = new Sprite();
			s.graphics.beginFill(0xff0000);
			s.graphics.drawCircle(5, 5, 5);
			s.graphics.endFill();
			var bmdSmall : BitmapData = new BitmapData(10, 10);
			bmdSmall.draw(s);
			var bmdLarge : BitmapData = new BitmapData(100, 100);
			bmdLarge.draw(bmdSmall, new Matrix(10, 0, 0, 10, 0, 0));
			addChild(new Bitmap(bmdLarge));
```
![image](https://user-images.githubusercontent.com/6400804/156356555-1d3107d6-a1a7-4caf-8d70-fb6fcb3dddad.png)

**Reply by Bimeche** [reply]
I searched a bit more, and found out that when switching to AIR 33, we also switched profiles to desktop (I checked since you explicitly said it in your answer, and we'll keep it that way from now on).
So, after finding this I tried your test, and got the same results as you did, with AIR 33 and AIR 29, when using desktop profile.
Sorry for the confusion and the time loss.

So in the end, the solution you gave with `<allowLowQuality>` would benefit us, as we could once again use the LOW and MEDIUM quality in our project while having the desktop profile.

**Comment by toysonlocation** [comment]
There have been a number of times back in Adobe days, and I expect the time will come again, when I want to do enlarged pixel interfaces. Being able to set the stage quality to low was how I used to do it. Is there another way?

**Reply by ajwfrost** [reply]
If you've drawn something to a bitmap data object, then you can presumably scale this up with `BitmapData.draw()` with an enlargement matrix and a 'false' value for smoothing?

**Comment by KupoGames** [comment]
If I recall correctly, Adobe said they disabled the LOW and MEDIUM options because they caused some rendering issues when an AIR app used multiple windows, or something like that. Either way, it was a feature that a lot of web game developers really wanted to see when porting their games to desktop! I've personally been using the web Flashplayer in my AIR apps to get around this limitation, which has been a pain, but the performance increase was worth it. I'm really happy to see this feature coming back.
